xds/src/test/java/io/grpc/xds/internal/certprovider/FileWatcherCertificateProviderTest.java

```diff
@@ -0,0 +1,350 @@
+
+  @Mock private CertificateProvider.Watcher mockWatcher;
+  @Mock private ScheduledExecutorService timeService;
+  @Mock private TimeProvider timeProvider;
+
+  @Rule public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  private String certFile;
+  private String keyFile;
+  private String rootFile;
+
+  private FileWatcherCertificateProvider provider;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+
+    DistributorWatcher watcher = new DistributorWatcher();
+    watcher.addWatcher(mockWatcher);
+
+    certFile = new File(tempFolder.getRoot(), CERT_FILE).getAbsolutePath();
+    keyFile = new File(tempFolder.getRoot(), KEY_FILE).getAbsolutePath();
+    rootFile = new File(tempFolder.getRoot(), ROOT_FILE).getAbsolutePath();
+    provider =
+        new FileWatcherCertificateProvider(
+            watcher, true, certFile, keyFile, rootFile, 600L, timeService, timeProvider);
+  }
+
+  private void populateTarget(
+      String certFileSource,
+      String keyFileSource,
+      String rootFileSource,
+      boolean deleteCurCert,
+      boolean deleteCurKey,
+      boolean deleteCurRoot)
+      throws IOException {
+    if (deleteCurCert) {
+      Files.delete(Paths.get(certFile));
+    }
+    if (certFileSource != null) {
+      certFileSource = CommonTlsContextTestsUtil.getTempFileNameForResourcesFile(certFileSource);
+      Files.copy(Paths.get(certFileSource), Paths.get(certFile), REPLACE_EXISTING);
+    }
+    if (deleteCurKey) {
+      Files.delete(Paths.get(keyFile));
+    }
+    if (keyFileSource != null) {
+      keyFileSource = CommonTlsContextTestsUtil.getTempFileNameForResourcesFile(keyFileSource);
+      Files.copy(Paths.get(keyFileSource), Paths.get(keyFile), REPLACE_EXISTING);
+    }
+    if (deleteCurRoot) {
+      Files.delete(Paths.get(rootFile));
+    }
+    if (rootFileSource != null) {
+      rootFileSource = CommonTlsContextTestsUtil.getTempFileNameForResourcesFile(rootFileSource);
+      Files.copy(Paths.get(rootFileSource), Paths.get(rootFile), REPLACE_EXISTING);
+    }
+  }
+
+  @Test
+  public void getCertificateAndCheckUpdates()
+      throws IOException, CertificateException, InterruptedException {
+    MeshCaCertificateProviderTest.TestScheduledFuture<?> scheduledFuture =
+        new MeshCaCertificateProviderTest.TestScheduledFuture<>();
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    populateTarget(CLIENT_PEM_FILE, CLIENT_KEY_FILE, CA_PEM_FILE, false, false, false);
+    provider.checkAndReloadCertificates();
+    verifyWatcherUpdates(CLIENT_PEM_FILE, CA_PEM_FILE);
+    verifyTimeServiceAndScheduledHandle();
+
+    reset(mockWatcher, timeService);
+    doReturn(scheduledFuture)
+        .when(timeService)
+        .schedule(any(Runnable.class), any(Long.TYPE), eq(TimeUnit.SECONDS));
+    provider.scheduledHandle.cancel();
```

Comment:
I'm not wild about calling an essentially private attribute in tests. This does not look like a real usecase. Why calling `cancel()` followed by `checkAndReloadCertificates()`?

Reply:
That's right, there is no use case for calling `cancel()` since it is not really exercised in `checkAndReloadCertificates()`. Will remove it